File: client-ios-swift/Messaging/Entities/Events/MessengerEvent.swift
/*
*  Copyright (c) 2011-2019, Zingaya, Inc. All rights reserved.
*/

import Foundation

enum MessengerEvent {
    case message (MessageEvent)
    case conversation (ConversationEvent)
    case service (ServiceEvent)
    case user (UserEvent)
    
    func either(isMessageEvent: ((MessageEvent) -> Void)? = nil,
                isConversationEvent: ((ConversationEvent) -> Void)? = nil,
                isServiceEvent: ((ServiceEvent) -> Void)? = nil,
                isUserEvent: ((UserEvent) -> Void)? = nil) {
        switch self {
        case let .message(messageEvent):
            isMessageEvent?(messageEvent)
        case let .conversation(conversationEvent):
            isConversationEvent?(conversationEvent)
        case let .service(serviceEvent):
            isServiceEvent?(serviceEvent)
        case let .user(userEvent):
            isUserEvent?(userEvent)
        }
    }
    
    var initiator: User {
        switch self {
        case let .message(messageEvent):
            return messageEvent.initiator
        case let .conversation(conversationEvent):
            return conversationEvent.initiator
        case let .service(serviceEvent):
            return serviceEvent.initiator
        case let .user(userEvent):
            return userEvent.initiator
        }
    }
    
    var sequence: Int {
        switch self {
        case let .message(messageEvent):
            return messageEvent.sequence
        case let .conversation(conversationEvent):
            return conversationEvent.sequence
        case let .service(serviceEvent):
            return serviceEvent.sequence
        case .user :
            return 0
        }
    }
}

File: client-android-kotlin/messaging/src/main/java/com/voximplant/demos/messaging/entity/events/MessengerEvent.kt
package com.voximplant.demos.messaging.entity.events

import androidx.room.Embedded
import androidx.room.Entity
import com.voximplant.demos.messaging.entity.Message

sealed class MessengerEvent {

    abstract val initiatorImId: Long
    abstract val sequence: Long

    @Entity(primaryKeys = ["sequence", "conversation"])
    data class ConversationEvent(
        override val initiatorImId: Long,
        val action: ConversationEventAction,
        override val sequence: Long,
        val conversation: String,
        val timestamp: Long
    ) : MessengerEvent()

    @Entity(primaryKeys = ["sequence", "conversation"])
    data class MessageEvent(
        override val initiatorImId: Long,
        var action: MessageEventAction,
        override val sequence: Long,
        @Embedded var message: Message,
        val timestamp: Long
    ) : MessengerEvent()

    data class ServiceEvent(
        override val initiatorImId: Long,
        val action: ServiceEventAction,
        override val sequence: Long,
        val conversationUUID: String
    ) : MessengerEvent()

    data class UserEvent(
        override val initiatorImId: Long,
        val action: UserEventAction
    ) : MessengerEvent() {
        override val sequence: Long
            get() = 0
    }

    fun either(
        isConversationEvent: ((ConversationEvent) -> Unit)? = null,
        isMessageEvent: ((MessageEvent) -> Unit)? = null,
        isServiceEvent: ((ServiceEvent) -> Unit)? = null,
        isUserEvent: ((UserEvent) -> Unit)? = null
    ) {
        when (this) {
            is ConversationEvent -> isConversationEvent?.invoke(this)
            is MessageEvent -> isMessageEvent?.invoke(this)
            is ServiceEvent -> isServiceEvent?.invoke(this)
            is UserEvent -> isUserEvent?.invoke(this)
        }
    }
}
